base/file.jl

```diff
@@ -493,40 +493,46 @@ const temp_prefix = "jl_"
 
 if Sys.iswindows()
 
-function _win_tempname(temppath::AbstractString, uunique::UInt32)
-    tempp = cwstring(temppath)
-    temppfx = cwstring(temp_prefix)
-    tname = Vector{UInt16}(undef, 32767)
-    uunique = ccall(:GetTempFileNameW, stdcall, UInt32,
-                    (Ptr{UInt16}, Ptr{UInt16}, UInt32, Ptr{UInt16}),
-                    tempp, temppfx, uunique, tname)
-    windowserror("GetTempFileName", uunique == 0)
-    lentname = something(findfirst(iszero, tname))
-    @assert lentname > 0
-    resize!(tname, lentname - 1)
-    return transcode(String, tname)
+# GUID  win32 api (consider # TODO: refactor)
+struct GUID
+    l1::Culong
+    w1::Cushort
+    w2::Cushort
+    b::NTuple{8,Cuchar}
+end
+
+function _tempname()
+    RPC_S_UUID_LOCAL_ONLY = 1824
+    RPC_S_UUID_NO_ADDRESS = 1739
+    RPC_S_OUT_OF_MEMORY = 14
+
+    id = Ref{GUID}()
+    r = ccall((:UuidCreate,:Rpcrt4), stdcall, Cint, (Ref{GUID},), id)
+    (r == 0 || r == RPC_S_UUID_LOCAL_ONLY) || (r == RPC_S_UUID_NO_ADDRESS && error("Cannot get Ethernet or token-ring hardware address for this computer."))
+
+    nameptr = Ref{Ptr{Cwchar_t}}()
+    r = ccall((:UuidToStringW, :Rpcrt4), stdcall, Cint, (Ref{GUID}, Ref{Ptr{Cwchar_t}}), id, nameptr)
+    r == 0 || (r == RPC_S_OUT_OF_MEMORY && error("Out of memory")) || error("cannot convert UIUD to string")
+
+    namebuf = unsafe_wrap(Vector{Cwchar_t}, nameptr[], ccall(:wcslen, UInt, (Ptr{Cwchar_t},), nameptr[]))
+    name = transcode(String, namebuf)
+
+    ccall((:RpcStringFreeW, :Rpcrt4), stdcall, Cint, (Ref{Ptr{Cwchar_t}},), nameptr) == 0
+
+    return name
 end
 
 function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
-    filename = _win_tempname(parent, UInt32(0))
+    filename = joinpath(parent, temp_prefix * _tempname())
     cleanup && temp_cleanup_later(filename)
-    return (filename, Base.open(filename, "r+"))
+    return (filename, Base.open(filename, "w+"))
 end
```

Comment:
Wouldn't this be simpler (but equivalent) as:
```suggestion
function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
     filename = tempname(parent; cleanup=false)
     io = Base.open(filename, "w+")
     cleanup && temp_cleanup_later(filename)
     return (filename, io)
 end
```